python/lsst/ts/watcher/rules/cp_verify_alarm.py

```diff
@@ -0,0 +1,296 @@
+                    # at least one type of test
+                    break
+
+        # For at least one type of test, if the majority of tests fail in
+        # the majority of detectors and the majority of exposures,
+        # then don't certify the calibration
+        if failed_exposures_counter >= max_number_failed_exposures:
+            certify_calib = False
+
+        # Return a dictionary with the thresholds to report
+        # them if verification fails.
+        thresholds = {
+            "MAX_FAILURES_PER_DETECTOR_PER_TEST_TYPE_THRESHOLD": max_number_failures_per_detector_per_test,
+            "MAX_FAILED_DETECTORS_THRESHOLD": max_number_failed_detectors,
+            "MAX_FAILED_TESTS_PER_EXPOSURE_THRESHOLD": failure_threshold_exposure,
+            "MAX_FAILED_EXPOSURES_THRESHOLD": max_number_failed_exposures,
+            "FINAL_NUMBER_OF_FAILED_EXPOSURES": failed_exposures_counter,
+        }
+
+        return certify_calib, total_counter_failed_tests, thresholds
+
+    def get_cp_verify_stats(self, response_verify):
+        """Get cp_verify statistics from the butler.
+
+        Parameters
+        ----------
+        response_verify : `dict`
+            OCPS call response.
+
+        Returns
+        -------
+        verify_stats : `dict`
+            Statistics from cp_verify.
+        """
+        job_id_verify = response_verify["job_id"]
+        # Loop over the entries of the 'results' list
+        # and look for the adecuate dataset type.
+        for entry in response_verify["results"]:
+            if "verifyBiasStats" in entry["uri"]:
+                verify_stats_string = "verifyBiasStats"
+                break
+            elif "verifyDarkStats" in entry["uri"]:
+                verify_stats_string = "verifyDarkStats"
+                break
+            elif "verifyFlatStats" in entry["uri"]:
+                verify_stats_string = "verifyFlatStats"
+                break
+            else:
+                # This is not a response from cp_verify
+                # bias, dark, or flat.
+                raise salobj.ExpectedError(
+                    f"Job {job_id_verify} is not a recognizable cp_verify run."
+                )
+
+        # Find repo and instrument
+        if len(response_verify["parameters"]["environment"]):
+            raise salobj.ExpectedError(
+                "response_verify['parameters']['environment'] is empty."
+            )
+
+        for entry in response_verify["parameters"]["environment"]:
+            if entry["name"] == "BUTLER_REPO":
+                repo = entry["value"]
+                instrument_name = repo.split("/")[-1]
+                break
+            else:
+                raise salobj.ExpectedError("OCPS response does not list a repo.")
+
+        # Collection name containing the verification outputs.
+        verify_collection = f"u/ocps/{job_id_verify}"
+        butler = dafButler.Butler(repo, collections=[verify_collection])
```

Comment:
If creating butlers is still expensive, please cache the ones you create. Specifically I suggest:
* For the cache use a dict of key: butler, where the key is `(repo, verify_collection)` or `(repo, job_id_verify)`, as you prefer. Together those seem to uniquely identify a butler.
* Create the empty cache in the `__init__` method.
* Empty the cache (and close all butlers) in the `stop` method. You probably know this, but this is a nice way to clear a dict:

```
while mydict:
    key, value = mydict.popitem()
    # ... do something with value, such as close it
```

Reply:
I'm not sure if this call will be expensive. A priori, it's just one butler call per calibration type (currently `BIAS`, `DARK`, and `FLAT`) to get the verification statistics dictionary.